### ISSUE TITLE
Cached downloads confuse multirust after deleting toolchain

### ISSUE THREAD

**Issue by huonw** [brson/multirust#60]
If one installs a toolchain and then immediately removes the toolchain directory, e.g.

```
$ multirust default nightly-2015-04-28
multirust: installing toolchain 'nightly-2015-04-28'
...
install: installing component 'rustc'
install: installing component 'cargo'
install: installing component 'rust-docs'

    Rust is ready to roll.

multirust: default toolchain set to 'nightly-2015-04-28'
$ rm -rf .multirust/toolchains/nightly-2015-04-28/
```

then, (obviously) rustc won't work:

```
$ rustc -V
multirust: toolchain 'nightly-2015-04-28' not installed. run `multirust update nightly-2015-04-28` to install
```

but running those commands doesn't help:

```
$ multirust update nightly-2015-04-28                                                                              
multirust: installing toolchain 'nightly-2015-04-28'
rustup: gpg available. signatures will be verified
rustup: downloading manifest for 'nightly-2015-04-28'
rustup: downloading toolchain for 'nightly-2015-04-28'
rustup: 'nightly-2015-04-28' is already up to date

$ rustc -V
multirust: toolchain 'nightly-2015-04-28' not installed. run `multirust update nightly-2015-04-28` to install
```

I believe this is because https://github.com/rust-lang/rustup/blob/404d1429dbb6aa46daf92ebf268c24a93474e136/rustup.sh#L1082 shortcircuits if only the download/check isn't necessary, but the return value of 20 is taken to be a short-circuit of the whole installation: https://github.com/rust-lang/rustup/blob/404d1429dbb6aa46daf92ebf268c24a93474e136/rustup.sh#L604-L608

**Comment by brson** [comment]
Yeah, that 'update-hash' is supposed to represent 'the thing that is installed' so must be kept in sync with what's actually there. If you delete the directory instead of calling `multirust remove-toolchain`, the metadata gets out of sync.

Still, I think this should work and users should expect to be able to reasonably futz with the the .multirust directory without invalidating its state.

One simple thing that might fix this specific scenario is to just check whether the directory exists, and if not deleting the update hash. Storing the update hash _inside_ the install directory would solve this particular scenario too, but in other cases I want to store less in this directory, so I don't know that I want to do that. Beyond that I'd hate to have to hash the entire directory to know whether the update hash is valid. Could hash just the rustc bin maybe.

Maybe there's a more reliable way to do this short-circuiting that I haven't thought of.

**Comment by huonw** [comment]
Checking whether the directory exists seems like a strict incremental improvement over just checking whether there's a cached download, even if it doesn't catch all cases when an installation may be invalid. 

FWIW, I think I actually saw this error the first time by ^C-ing `multirust default nightly-2015-...` at a bad time, meaning it may be able to be triggered without manual `rm`s. (I'm not 100% sure that this was the problem.)

I find manually `rm`ing particularly useful with toolchains because I install nightlies by date to allow easy bisection/testing with older versions (without having to redownload), so I collect gigabytes and gigabytes of toolchains, and `rm`ing weeks at a time with a glob is nice.
